src/ILCompiler.Compiler/src/Compiler/DevirtualizationManager.cs

```diff
@@ -54,7 +54,7 @@ public virtual bool IsEffectivelySealed(MethodDesc method)
         /// Note that if <paramref name="implType"/> is a value type, the result of the resolution
         /// might have to be treated as an unboxing thunk by the caller.
         /// </remarks>
-        public MethodDesc ResolveVirtualMethod(MethodDesc declMethod, TypeDesc implType)
+        public virtual MethodDesc ResolveVirtualMethod(MethodDesc declMethod, TypeDesc implType)
```

Comment:
Will this actually do the right thing when devirtualizing methods on arrays (e.g. calling IStructuralEquatable.GetHashCode on an array type)? Seems like the override you're adding will directly call into algorithms that expect `implType` to be a `MetadataType`. This probably won't do the right thing if `implType` is an array, which is the only extra thing this non-overridable method does on top of the overridable method 13 lines below.

I think the actual problem is just that `GetClosestDefType` called on line 67 doesn't do the right thing for arrays when its virtual methods are not backed by `Array<T>`. I think `GetClosestDefType` should just return System.Array on R2R.

Your current solution doesn't hit a problem because it ends up rejecting devirtualizing arrays due to the version bubble crossing (unless there's e.g. an IStructuralEquatable.GetHashCode call on an array within CoreLib), but it will be a problem with a big version bubble. I would rather not introduce this extension point because it's not going to survive long term anyway. Let's fix `GetClosestDefType` to return System.Array if it can't find `Array<T>`.

Reply:
I have reverted this bit in 8th commit. I think this was just caused by my limited understanding at the time I was implementing the CPAOT devirtualization manager - we just need to override the proper internal method to perform the cross-bubble checks.